maintainers/scripts/update.py

```diff
@@ -31,15 +40,35 @@ def main(max_workers, keep_going, packages):
         eprint()
         eprint('Running update for:')
 
-        with concurrent.futures.ProcessPoolExecutor(max_workers=max_workers) as executor:
+        with contextlib.ExitStack() as stack, concurrent.futures.ThreadPoolExecutor(max_workers=max_workers, thread_name_prefix=thread_name_prefix) as executor:
+            global temp_dirs
+
+            if commit:
+                temp_dirs = {f'{thread_name_prefix}_{str(i)}': (stack.enter_context(tempfile.TemporaryDirectory()), threading.Lock()) for i in range(max_workers)}
+
+                for wt, _lock in temp_dirs.values():
+                    subprocess.run(['git', 'worktree', 'add', wt], check=True)
+            else:
+                toplevel = subprocess.run(['git', 'rev-parse', '--show-toplevel'], stdout=subprocess.PIPE, check=True).stdout.decode('utf-8').strip()
+                temp_dirs = {f'{thread_name_prefix}_{str(i)}': (toplevel, None) for i in range(max_workers)}
+
             for package in packages:
-                updates[executor.submit(run_update_script, package)] = package
+                updates[executor.submit(run_update_script, package, commit)] = package
 
             for future in concurrent.futures.as_completed(updates):
                 package = updates[future]
 
                 try:
-                    future.result()
+                    p = future.result()
+                    if commit and 'commit' in package['supportedFeatures']:
+                        thread_name = package['thread']
+                        worktree, lock = temp_dirs[thread_name]
+                        changes = json.loads(p.stdout)
+                        for change in changes:
```

Comment:
too much nesting. Consider a small separate function for the following lines

Reply:
I would say a much more pressing issue is how tightly coupled is everything with console input/output and the interruption not working very well.

Something like this looks better but interruption is still not handled gracefully. Perhaps using `queue.Queue` would make it nicer.

```patch
--- a/maintainers/scripts/update.py
+++ b/maintainers/scripts/update.py
@@ -9,6 +9,7 @@
 import threading
 
 updates = {}
+temp_dirs = {}
 
 thread_name_prefix='UpdateScriptThread'
 
@@ -28,6 +29,41 @@
 
     return subprocess.run(package['updateScript'], stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True, cwd=worktree)
 
+def update_packages(packages, max_workers, commit):
+    with contextlib.ExitStack() as stack, concurrent.futures.ThreadPoolExecutor(max_workers=max_workers, thread_name_prefix=thread_name_prefix) as executor:
+
+        if commit:
+            for i in range(max_workers):
+                temp_dirs[f'{thread_name_prefix}_{str(i)}'] = stack.enter_context(tempfile.TemporaryDirectory()), threading.Lock()
+
+            for wt, _lock in temp_dirs.values():
+                subprocess.run(['git', 'worktree', 'add', wt], check=True)
+
+        for package in packages:
+            updates[executor.submit(run_update_script, package, commit)] = package
+
+        for future in concurrent.futures.as_completed(updates):
+            package = updates[future]
+
+            try:
+                p = future.result()
+                if commit and 'commit' in package['supportedFeatures']:
+                    thread_name = package['thread']
+                    worktree, lock = temp_dirs[thread_name]
+                    changes = json.loads(p.stdout)
+                    for change in changes:
+                        subprocess.run(['git', 'add'] + change['files'], check=True, cwd=worktree)
+                        commit_message = '{attrName}: {oldVersion} → {newVersion}'.format(**change)
+                        subprocess.run(['git', 'commit', '-m', commit_message], check=True, cwd=worktree)
+                        subprocess.run(['git', 'cherry-pick', os.path.basename(worktree)], check=True)
+                yield package, True, None
+            except subprocess.CalledProcessError as e:
+                yield package, False, e.stdout
+            finally:
+                if commit and 'commit' in package['supportedFeatures']:
+                    lock.release()
+
+
 def main(max_workers, keep_going, commit, packages):
     with open(sys.argv[1]) as f:
         packages = json.load(f)
@@ -43,49 +79,22 @@
         eprint()
         eprint('Running update for:')
 
-        with contextlib.ExitStack() as stack, concurrent.futures.ThreadPoolExecutor(max_workers=max_workers, thread_name_prefix=thread_name_prefix) as executor:
-            global temp_dirs
-
-            if commit:
-                temp_dirs = {f'{thread_name_prefix}_{str(i)}': (stack.enter_context(tempfile.TemporaryDirectory()), threading.Lock()) for i in range(max_workers)}
-
-                for wt, _lock in temp_dirs.values():
-                    subprocess.run(['git', 'worktree', 'add', wt], check=True)
-
-            for package in packages:
-                updates[executor.submit(run_update_script, package, commit)] = package
-
-            for future in concurrent.futures.as_completed(updates):
-                package = updates[future]
-
-                try:
-                    p = future.result()
-                    if commit and 'commit' in package['supportedFeatures']:
-                        thread_name = package['thread']
-                        worktree, lock = temp_dirs[thread_name]
-                        changes = json.loads(p.stdout)
-                        for change in changes:
-                            subprocess.run(['git', 'add'] + change['files'], check=True, cwd=worktree)
-                            commit_message = '{attrName}: {oldVersion} → {newVersion}'.format(**change)
-                            subprocess.run(['git', 'commit', '-m', commit_message], check=True, cwd=worktree)
-                            subprocess.run(['git', 'cherry-pick', os.path.basename(worktree)], check=True)
-                    eprint(f" - {package['name']}: DONE.")
-                except subprocess.CalledProcessError as e:
-                    eprint(f" - {package['name']}: ERROR")
-                    eprint()
-                    eprint(f"--- SHOWING ERROR LOG FOR {package['name']} ----------------------")
-                    eprint()
-                    eprint(e.stdout.decode('utf-8'))
-                    with open(f"{package['pname']}.log", 'wb') as f:
-                        f.write(e.stdout)
-                    eprint()
-                    eprint(f"--- SHOWING ERROR LOG FOR {package['name']} ----------------------")
-
-                    if not keep_going:
-                        sys.exit(1)
-                finally:
-                    if commit and 'commit' in package['supportedFeatures']:
-                        lock.release()
+        for package, status, output in update_packages(packages, max_workers, commit):
+            if package['status']:
+                eprint(f" - {package['name']}: DONE.")
+            else:
+                eprint(f" - {package['name']}: ERROR")
+                eprint()
+                eprint(f"--- SHOWING ERROR LOG FOR {package['name']} ----------------------")
+                eprint()
+                eprint(output.decode('utf-8'))
+                with open(f"{package['pname']}.log", 'wb') as f:
+                    f.write(output)
+                eprint()
+                eprint(f"--- SHOWING ERROR LOG FOR {package['name']} ----------------------")
+
+                if not keep_going:
+                    sys.exit(1)
 
         eprint()
         eprint('Packages updated!')
```

Reply:
I moved it to separate function. I am also thinking about switching from `ThreadPoolExecutor` to asyncio and https://pypi.org/project/asyncio-pool/